### PR TITLE
JP-2375: NIRISS V3 of CONF files

### LINKED ISSUE #6558: Update NIRISS WFSS transforms to use V3 of the CONF files
_Issue [JP-2375](https://jira.stsci.edu/browse/JP-2375) was created on JIRA by [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva):_

Version 3 of the NIRISS CONF files implements a quadratic polynomial in terms of an independent variable "t" , where the coefficients are polynomials.

spacetelescope/jwreftools#42

implements the translation of the CONF files to reference files of reftype "specwcs" and "wavelengthrange" in ASDF format.

The assign_wcs transforms need to be modified to use these new polynomials.

Currently the transform is
```java
dx = cpoly_0 + t* cpoly_1
```
The new transform should be 
```java
dx = cpoly_0 + t* cpoly_1 + t**2 * cpoly_2
```
where "cpoly_0", "cpoly_1" and "cpoly_2" are polynomials defined in the "specwcs" reference file.

In addition, we need a decision on how to define the inverse transform.

### PR DESCRIPTION
**Pull request by nden**
Closes #6558
Resolves [JP-2375](https://jira.stsci.edu/browse/JP-2375)

Checklist
- [x] Tests
- [ ] Documentation
- [x] Change log
- [x] Milestone
- [x] Label(s)

### CODE REVIEW

File: jwst/assign_wcs/tests/test_niriss.py [hunk 1]
```diff
@@ -125,12 +125,14 @@ def traverse_wfss_trace(filtername):
     assert orderdet == orderin
 
 
+@pytest.mark.xfail(reason="New specwcs reference files are needed in CRDS")
```
Comment (hbushouse) [inline jwst/assign_wcs/tests/test_niriss.py:128]: Is this still true or are the files recently delivered by Swara sufficient?

File: jwst/background/tests/test_background.py [hunk 2]
```diff
@@ -244,6 +244,7 @@ def test_nrc_wfss_background(filters, pupils, detectors, make_wfss_datamodel):
         assert np.isclose([pipeline_reference_mean], [test_reference_mean], rtol=1e-1)
 
 
+@pytest.mark.xfail(reason="Needs new reference specawcs reference files in CRDS")
```
Comment (hbushouse) [inline jwst/background/tests/test_background.py:247]: Still true?

File: jwst/extract_2d/grisms.py [hunk 3]
```diff
@@ -411,6 +411,7 @@ def extract_grism_objects(input_model,
             # The bounding boxes here are also limited to the size of the detector
             # The check for boxes entirely off the detector is done in create_grism_bbox right now
             y, x = obj.order_bounding[order]
+            log.info(f'YYY, {y}, {clamp(y[0], 0, input_model.meta.subarray.ysize)}')
```
Comment (hbushouse) [inline jwst/extract_2d/grisms.py:414]: Is this the type of info we want appearing in a category `info` message, or is it better suited to `debug` mode?

File: jwst/extract_2d/grisms.py [hunk 4]
```diff
@@ -445,20 +446,24 @@ def extract_grism_objects(input_model,
...
                                                  order_model) | tr
-
-                ext_data = input_model.data[ymin: ymax + 1, xmin: xmax + 1].copy()
-                ext_err = input_model.err[ymin: ymax + 1, xmin: xmax + 1].copy()
-                ext_dq = input_model.dq[ymin: ymax + 1, xmin: xmax + 1].copy()
+                from gwcs.utils import _toindex
+                y_slice = slice(_toindex(ymin), _toindex(ymax))
+                x_slice = slice(_toindex(xmin), _toindex(xmax))
+                ext_data = input_model.data[y_slice, x_slice].copy()
+                # ext_err = input_model.err[ymin: ymax + 1, xmin: xmax + 1].copy()
```
Comment (hbushouse) [inline jwst/extract_2d/grisms.py:453]: Is there a reason to leave the commented lines in here?

File: jwst/transforms/models.py [hunk 5]
```diff
@@ -1147,15 +1148,16 @@ def evaluate(self, x, y, x0, y0, order):
...
         lmodel = self.lmodels[iorder]
-        dx = xmodel[0](x00, y00) + t * xmodel[1](x00, y00)
-        dy = ymodel[0](x00, y00) + t * ymodel[1](x00, y00)
+        dx = xmodel[0](x00, y00) + t * xmodel[1](x00, y00) + t**2 * xmodel[2](x00, y00)
+        dy = ymodel[0](x00, y00) + t * ymodel[1](x00, y00) + t**2 * ymodel[2](x00, y00)
 
         if self.theta != 0.0:
             rotate = Rotation2D(self.theta)
             dx, dy = rotate(dx, dy)
+        # so = np.argsort(dy)
```
Comment (hbushouse) [inline jwst/transforms/models.py:1157]: These commented lines appear to be duplicates of the active lines below.

**Review by hbushouse — COMMENTED** [review]
Did Swara verify the functionality of the PR branch code?

**Comment by hbushouse** [comment]
I've tested this with in-flight NIRISS grism observations and a patched `wfss_contam` step. It all seems to be working properly. So I would be OK with merging this once my review questions have been answered.

File: jwst/assign_wcs/tests/test_niriss.py (hunk 1 above)
Comment (nden) [reply jwst/assign_wcs/tests/test_niriss.py]: removed

File: jwst/background/tests/test_background.py (hunk 2 above)
Comment (nden) [reply jwst/background/tests/test_background.py:247]: removed

File: jwst/extract_2d/grisms.py (hunk 3 above)
Comment (nden) [reply jwst/extract_2d/grisms.py]: changed

File: jwst/extract_2d/grisms.py (hunk 4 above)
Comment (nden) [reply jwst/extract_2d/grisms.py]: no

File: jwst/transforms/models.py (hunk 5 above)
Comment (nden) [reply jwst/transforms/models.py]: removed

**Comment by nden** [comment]
I addressed the comments and enabled previously xfailed tests. Swara tested with this PR and hit the error in the contamination step.

**Comment by hbushouse** [comment]
Whoops - you've got 1 little flake8 issue.

**Comment by hbushouse** [comment]
Did a full regression test ever get run against this PR branch?  A couple of us tested it on NIRISS WFSS processing only, which seemed to run OK, but never a full regtest.

File: jwst/extract_2d/grisms.py [hunk 6]
```diff
@@ -445,20 +446,22 @@ def extract_grism_objects(input_model,
                                                  xcenter_model &
                                                  ycenter_model &
                                                  order_model) | tr
-
-                ext_data = input_model.data[ymin: ymax + 1, xmin: xmax + 1].copy()
-                ext_err = input_model.err[ymin: ymax + 1, xmin: xmax + 1].copy()
-                ext_dq = input_model.dq[ymin: ymax + 1, xmin: xmax + 1].copy()
+                from gwcs.utils import _toindex
+                y_slice = slice(_toindex(ymin), _toindex(ymax))
```
Comment (hbushouse) [inline jwst/extract_2d/grisms.py:450]: In retrospect, I think this is a mistake and causes the actual number of extracted rows/cols to be short by 1. For example, if you have ymin=941.8 and ymax=951.8, the `_toindex` function converts those limits to 942 and 952. Those are in fact the *inclusive* limits of what we want to extract, i.e. a total of 11 rows, 942 through 952, inclusive. But when those are used directly as python slice indexes, you of course end up extracting only 942 through 951, a total of 10 rows. That's why the old lines of code explicitly added 1 to the ymax/xmax values before using them in the python slicing operation. This is leading to extraction boxes being undersized by 1 row/col.

Comment (hbushouse) [reply jwst/extract_2d/grisms.py:450]: I believe it's necessary to again add 1 to `_toindex(ymax)` and `_toindex(xmax)` when constructing `y_slice` and `x_slice`.

**Pull request merged**